### PR TITLE
[REVIEW] Java standard deviation reduction

### PR DESCRIPTION
**Pull request by rwlee**
Add java support for sample standard deviation column reduction operation. Restructures java reduction unit testing to do element wise comparison of the scalar class, allowing for assertion checks to include a delta value for slight differences in standard deviation floating point result. Standard deviation return value is forced to FLOAT64 unless the input column is FLOAT32 or FLOAT32 is specified in the call.

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 1]
```diff
@@ -1091,6 +1111,8 @@ public Scalar reduce(ReductionOp op) {
    * {@link Scalar#isValid()} method of the result will return false.
    */
   public Scalar reduce(ReductionOp op, DType outType) {
+    if(op == ReductionOp.STD && outType != DType.FLOAT32)
+      outType = DType.FLOAT64;
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java]: If they passed in a specific output type we should honor it.  If it is not going to work, then we should have an assertion.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 2]
```diff
@@ -1064,6 +1064,26 @@ public Scalar sumOfSquares(DType outType) {
...
   }
 
+  /**
+   * Returns the sample standard deviation of all values in the column,
+   * returning a scalar of the same type as this column. Null's are not
+   * counted as an element of the column when calculating the standard
+   * deviation. Output type is FLOAT64 unless input column is FLOAT32.
+   */
+  public Scalar standardDeviation() {
+    return standardDeviation(type);
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java]: This is the proper place to set a default output type.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java (hunk 1 above)
Comment (rwlee) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java]: So a user using this direct reduce method (over the standardDeviation method) directly that does not specify a valid type should receive an error from the java side before any actual execution?
```
    if(op == ReductionOp.STD)
      assert (outType == DType.FLOAT64 || outType == DType.FLOAT32);
```

Comment (revans2) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java]: If the native side returns an error message that a user could use to debug it, then I would prefer to have the native side handle it, but if the error message is not clear, or if it panics then we should have the assertion.

Comment (rwlee) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java]: Native error is useful and informative enough to debug it.

Abbreviated error trace
```
ai.rapids.cudf.CudfException: cuDF failure at: ../cudf/cpp/src/reductions/compound.cuh:114: Unsupported output data type
	at ai.rapids.cudf.Cudf.reduce(Native Method)
	at ai.rapids.cudf.Cudf.reduce(Cudf.java:121)
	at ai.rapids.cudf.ColumnVector.reduce(ColumnVector.java:1116)
	at ai.rapids.cudf.ColumnVector.reduce(ColumnVector.java:1100)
```

**Comment by jlowe** [comment]
rerun tests

**Comment by rwlee** [comment]
Tests need to be reran

**Comment by rwlee** [comment]
rerun tests

**Comment by jlowe** [comment]
@kkraus14 this looks like another PR being held hostage by the broken CI pipeline.  We can kick off the tests manually which shows a link to the test build.  However later it loses track of the build somehow.  The link disappears and leaves the PR in a state where it thinks the test build is still running yet has no record of it.

**Comment by kkraus14** [comment]
add to whitelist

**Comment by kkraus14** [comment]
@jlowe If that happens it's because the person isn't whitelisted for CI and CI doesn't run code without someone approving by commenting "ok to test" or an admin adding to the whitelist via "add to whitelist". This is to prevent running potentially malicious code in CI.

**Comment by jlowe** [comment]
@kkraus14 I think this is read to go.  Passes the builds but fails the test runs due to conda environment issues.  I believe the latter is a known issue with the CI test pipeline.

**Pull request merged**